src/passkey_child/passkey_child_common.c

```diff
@@ -319,3 +323,105 @@ register_key(struct passkey_data *data)
 
     return ret;
 }
+
+errno_t
+authenticate(struct passkey_data *data)
+{
+    fido_assert_t *assert = NULL;
+    fido_dev_info_t *dev_list = NULL;
+    fido_dev_t *dev = NULL;
+    size_t dev_number = 0;
+    errno_t ret;
+
+    dev_list = fido_dev_info_new(DEVLIST_SIZE);
+    if (dev_list == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = list_devices(dev_list, &dev_number);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    if (dev_number == 0) {
+        DEBUG(SSSDBG_OP_FAILURE, "No device found. Aborting.\n");
+        ret = ENOENT;
+        goto done;
+    }
+
+    assert = fido_assert_new();
+    if (assert == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_assert_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = prepare_assert(data, assert);
+    if (ret != FIDO_OK) {
+        goto done;
+    }
+
+    for (int i = 0; i < dev_number; i++) {
+        DEBUG(SSSDBG_TRACE_FUNC,
+              "Attempting authentication with device number %d.\n", i + 1);
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Decoding public key.\n");
+        ret = decode_public_key(data);
+        if (ret != FIDO_OK) {
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Obtaining device assertation.\n");
+        ret = get_device_assert(data, dev_list, dev_number, assert,
+                                &dev);
+        if (ret != FIDO_OK) {
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Comparing the device and policy options.\n");
+        ret = get_device_options(data, dev);
```

Comment:
This function mutates `data` and no copy of the original is kept. This will have unintended side-effects if there are more than one authenticator connected to the system and the outer loop runs more than once.

Reply:
I think this doesn't apply any longer.